In [88]:
#importing dependencies
import numpy as np
import pandas as pd

import warnings
warnings.filterwarnings("ignore")

In [89]:
#reading the dataset
laliga_df = pd.read_csv("Laliga.csv")
laliga_df

Pos              Team  Seasons Points GamesPlayed GamesWon GamesDrawn  \
0     1       Real Madrid       86   4385        2762     1647        552   
1     2         Barcelona       86   4262        2762     1581        573   
2     3   Atletico Madrid       80   3442        2614     1241        598   
3     4          Valencia       82   3386        2664     1187        616   
4     5   Athletic Bilbao       86   3368        2762     1209        633   
..  ...               ...      ...    ...         ...      ...        ...   
56   57             Xerez        1     34          38        8         10   
57   58            Condal        1     22          30        7          8   
58   59   Atletico Tetuan        1     19          30        7          5   
59   60  Cultural Leonesa        1     14          30        5          4   
60   61            Girona        1      -           -        -          -   

   GamesLost GoalsFor GoalsAgainst Champion Runner-up Third Fourth Fifth  \
0        563     5947         3140       33        23     8      8     3   
1        608     5900         3114       25        25    12     12     4   
2        775     4534         3309       10         8    16      9     7   
3        861     4398         3469        6         6    10     11    10   
4        920     4631         3700        8         7    10      5     8   
..       ...      ...          ...      ...       ...   ...    ...   ...   
56        20       38           66        -         -     -      -     -   
57        15       37           57        -         -     -      -     -   
58        18       51           85        -         -     -      -     -   
59        21       34           65        -         -     -      -     -   
60         -        -            -        -         -     -      -     -   

   Sixth   T    Debut Since/LastApp  BestPosition  
0      4  79     1929          1929             1  
1      6  83     1929          1929             1  
2      6  56     1929       2002-03             1  
3      7  50  1931-32       1987-88             1  
4     10  49     1929          1929             1  
..   ...  ..      ...           ...           ...  
56     -   -  2009-10       2009-10            20  
57     -   -  1956-57       1956-57            16  
58     -   -  1951-52       1951-52            16  
59     -   -  1955-56       1955-56            15  
60     -   -  2017-18       2017-18             9  

[61 rows x 20 columns]

In [90]:
#getting top 5 rows
laliga_df.head()

Pos             Team  Seasons Points GamesPlayed GamesWon GamesDrawn  \
0    1      Real Madrid       86   4385        2762     1647        552   
1    2        Barcelona       86   4262        2762     1581        573   
2    3  Atletico Madrid       80   3442        2614     1241        598   
3    4         Valencia       82   3386        2664     1187        616   
4    5  Athletic Bilbao       86   3368        2762     1209        633   

  GamesLost GoalsFor GoalsAgainst Champion Runner-up Third Fourth Fifth Sixth  \
0       563     5947         3140       33        23     8      8     3     4   
1       608     5900         3114       25        25    12     12     4     6   
2       775     4534         3309       10         8    16      9     7     6   
3       861     4398         3469        6         6    10     11    10     7   
4       920     4631         3700        8         7    10      5     8    10   

    T    Debut Since/LastApp  BestPosition  
0  79     1929          1929             1  
1  83     1929          1929             1  
2  56     1929       2002-03             1  
3  50  1931-32       1987-88             1  
4  49     1929          1929             1

In [91]:
#getting last 5 rows
laliga_df.tail()

Pos              Team  Seasons Points GamesPlayed GamesWon GamesDrawn  \
56   57             Xerez        1     34          38        8         10   
57   58            Condal        1     22          30        7          8   
58   59   Atletico Tetuan        1     19          30        7          5   
59   60  Cultural Leonesa        1     14          30        5          4   
60   61            Girona        1      -           -        -          -   

   GamesLost GoalsFor GoalsAgainst Champion Runner-up Third Fourth Fifth  \
56        20       38           66        -         -     -      -     -   
57        15       37           57        -         -     -      -     -   
58        18       51           85        -         -     -      -     -   
59        21       34           65        -         -     -      -     -   
60         -        -            -        -         -     -      -     -   

   Sixth  T    Debut Since/LastApp  BestPosition  
56     -  -  2009-10       2009-10            20  
57     -  -  1956-57       1956-57            16  
58     -  -  1951-52       1951-52            16  
59     -  -  1955-56       1955-56            15  
60     -  -  2017-18       2017-18             9

In [92]:
#checking statistical summary of dataset
laliga_df.describe()

Pos    Seasons  BestPosition
count  61.000000  61.000000     61.000000
mean   31.000000  24.000000      7.081967
std    17.752934  26.827225      5.276663
min     1.000000   1.000000      1.000000
25%    16.000000   4.000000      3.000000
50%    31.000000  12.000000      6.000000
75%    46.000000  38.000000     10.000000
max    61.000000  86.000000     20.000000

In [93]:
#checking datatypes of features and null values
laliga_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61 entries, 0 to 60
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Pos            61 non-null     int64 
 1   Team           61 non-null     object
 2   Seasons        61 non-null     int64 
 3   Points         61 non-null     object
 4   GamesPlayed    61 non-null     object
 5   GamesWon       61 non-null     object
 6   GamesDrawn     61 non-null     object
 7   GamesLost      61 non-null     object
 8   GoalsFor       61 non-null     object
 9   GoalsAgainst   61 non-null     object
 10  Champion       61 non-null     object
 11  Runner-up      61 non-null     object
 12  Third          61 non-null     object
 13  Fourth         61 non-null     object
 14  Fifth          61 non-null     object
 15  Sixth          61 non-null     object
 16  T              61 non-null     object
 17  Debut          61 non-null     object
 18  Since/LastApp  61 non-null     o

### Q1.Read the data set and replace dashes with 0 to make sure you can perform arithmetic operations on the data.

In [94]:
#replacing dash(-) with 0.
laliga_df = laliga_df.replace(to_replace="-",value=0)

In [95]:
laliga_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61 entries, 0 to 60
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Pos            61 non-null     int64 
 1   Team           61 non-null     object
 2   Seasons        61 non-null     int64 
 3   Points         61 non-null     object
 4   GamesPlayed    61 non-null     object
 5   GamesWon       61 non-null     object
 6   GamesDrawn     61 non-null     object
 7   GamesLost      61 non-null     object
 8   GoalsFor       61 non-null     object
 9   GoalsAgainst   61 non-null     object
 10  Champion       61 non-null     object
 11  Runner-up      61 non-null     object
 12  Third          61 non-null     object
 13  Fourth         61 non-null     object
 14  Fifth          61 non-null     object
 15  Sixth          61 non-null     object
 16  T              61 non-null     object
 17  Debut          61 non-null     object
 18  Since/LastApp  61 non-null     o

**Observations:**
* We need to change the datatypes of various columns as we can see that the datatype of columns are showing object but we have numerical data.
* So, we will first change the datatype so that its easy for us to perform operations.

In [96]:
#changing datatype from object to int
laliga_df[['Points','GamesPlayed','GamesWon','GamesDrawn','GamesLost','GoalsFor','GoalsAgainst','Champion','Runner-up','Third','Fourth','Fifth','Sixth','T']] = laliga_df[['Points','GamesPlayed','GamesWon','GamesDrawn','GamesLost','GoalsFor','GoalsAgainst','Champion','Runner-up','Third','Fourth','Fifth','Sixth','T']].apply(pd.to_numeric)
print(laliga_df.dtypes)

Pos               int64
Team             object
Seasons           int64
Points            int64
GamesPlayed       int64
GamesWon          int64
GamesDrawn        int64
GamesLost         int64
GoalsFor          int64
GoalsAgainst      int64
Champion          int64
Runner-up         int64
Third             int64
Fourth            int64
Fifth             int64
Sixth             int64
T                 int64
Debut            object
Since/LastApp    object
BestPosition      int64
dtype: object


### Q2.Print all the teams which have started playing between 1930-1980. 

In [97]:
# filter dataframe                                  
filtered_teams = np.where((laliga_df['Debut'] >= '1930') & (laliga_df['Debut'] <= '1980'))
print(filtered_teams)
display(laliga_df.loc[filtered_teams])

(array([ 3,  5,  8,  9, 10, 11, 12, 14, 15, 16, 17, 18, 19, 21, 22, 23, 25,
       26, 27, 28, 29, 30, 31, 32, 34, 37, 39, 40, 41, 46, 49, 50, 52, 54,
       57, 58, 59], dtype=int64),)


Pos                 Team  Seasons  Points  GamesPlayed  GamesWon  \
3     4             Valencia       82    3386         2664      1187   
5     6              Sevilla       73    2819         2408       990   
8     9             Zaragoza       58    2109         1986       698   
9    10           Real Betis       51    1884         1728       606   
10   11  Deportivo La Coruna       45    1814         1530       563   
11   12           Celta Vigo       51    1789         1698       586   
12   13           Valladolid       42    1471         1466       463   
14   15       Sporting Gijon       43    1389         1458       471   
15   16              Osasuna       37    1351         1318       426   
16   17               Malaga       36    1314         1255       390   
17   18               Oviedo       38    1174         1192       408   
18   19             Mallorca       27    1148          988       333   
19   20           Las Palmas       33    1020         1096       367   
21   22              Granada       23     667          742       218   
22   23       Rayo Vallecano       17     662          652       189   
23   24                Elche       21     606          678       203   
25   26             Hercules       20     538          628       184   
26   27             Tenerife       13     510          494       155   
27   28               Murcia       18     445          586       145   
28   29               Alaves       12     421          380       125   
29   30              Levante       11     416          402       113   
30   31            Salamanca       12     375          423       123   
31   32             Sabadell       14     353          426       129   
32   33                Cadiz       12     343          448       104   
34   35            Castellon       11     285          334       103   
37   38              Cordoba        9     230          282        82   
39   40           Recreativo        5     188          186        50   
40   41            Burgos CF        6     168          204        59   
41   42           Pontevedra        6     150          180        53   
46   47            Gimnastic        4      91          116        34   
49   50             Alcoyano        4      76          108        30   
50   51                 Jaen        3      71           90        29   
52   53           AD Almeria        2      52           68        17   
54   55               Lleida        2      40           68        13   
57   58               Condal        1      22           30         7   
58   59      Atletico Tetuan        1      19           30         7   
59   60     Cultural Leonesa        1      14           30         5   

    GamesDrawn  GamesLost  GoalsFor  GoalsAgainst  Champion  Runner-up  Third  \
3          616        861      4398          3469         6          6     10   
5          531        887      3680          3373         1          4      4   
8          522        766      2683          2847         0          1      4   
9          440        682      2159          2492         1          0      2   
10         392        575      2052          2188         1          5      4   
11         389        723      2278          2624         0          0      0   
12         384        619      1767          2180         0          0      0   
14         358        629      1753          2152         0          1      1   
15         327        565      1500          1834         0          0      0   
16         330        535      1421          1763         0          0      0   
17         292        492      1642          1951         0          0      3   
18         256        399      1182          1371         0          0      2   
19         242        487      1347          1746         0          1      1   
21         175        349       819          1157         0          0      0   
22         148        305       760          1088   

**Observations:**
* These are the Teams who have done Debut in between 1930 to 1980.

### Q3. Print the list of teams which came Top 5 in terms of points.

In [98]:
#Sorting Teams based on points
top5 = laliga_df.sort_values(['Points'], ascending=[False])
top5.head(5)

Pos             Team  Seasons  Points  GamesPlayed  GamesWon  GamesDrawn  \
0    1      Real Madrid       86    4385         2762      1647         552   
1    2        Barcelona       86    4262         2762      1581         573   
2    3  Atletico Madrid       80    3442         2614      1241         598   
3    4         Valencia       82    3386         2664      1187         616   
4    5  Athletic Bilbao       86    3368         2762      1209         633   

   GamesLost  GoalsFor  GoalsAgainst  Champion  Runner-up  Third  Fourth  \
0        563      5947          3140        33         23      8       8   
1        608      5900          3114        25         25     12      12   
2        775      4534          3309        10          8     16       9   
3        861      4398          3469         6          6     10      11   
4        920      4631          3700         8          7     10       5   

   Fifth  Sixth   T    Debut Since/LastApp  BestPosition  
0      3      4  79     1929          1929             1  
1      4      6  83     1929          1929             1  
2      7      6  56     1929       2002-03             1  
3     10      7  50  1931-32       1987-88             1  
4      8     10  49     1929          1929             1

**Observations:**
* **Real Madrid** is the top team in terms of points.

### Q4. Write a function with name “Goal_diff_count” which should return all the teams with their Goal Differences. Using the same function, find the team which has maximum and minimum goal difference. 

In [99]:
def goal_diff_count(laliga_df):
    laliga_df['goal_diff'] = laliga_df['GoalsFor'] - laliga_df['GoalsAgainst']
    print(laliga_df[['Team','goal_diff']])
    print("******************************************\n")
    
    print("Team with maximum Goal difference")
    print(laliga_df[laliga_df['goal_diff'] == laliga_df['goal_diff'].max()])
    
    print("******************************************\n")
    print("Team with minimum Goal difference")
    print(laliga_df[laliga_df['goal_diff'] == laliga_df['goal_diff'].min()])

goal_diff_count(laliga_df)

                Team  goal_diff
0        Real Madrid       2807
1          Barcelona       2786
2    Atletico Madrid       1225
3           Valencia        929
4    Athletic Bilbao        931
..               ...        ...
56             Xerez        -28
57            Condal        -20
58   Atletico Tetuan        -34
59  Cultural Leonesa        -31
60            Girona          0

[61 rows x 2 columns]
******************************************

Team with maximum Goal difference
   Pos         Team  Seasons  Points  GamesPlayed  GamesWon  GamesDrawn  \
0    1  Real Madrid       86    4385         2762      1647         552   

   GamesLost  GoalsFor  GoalsAgainst  ...  Runner-up  Third  Fourth  Fifth  \
0        563      5947          3140  ...         23      8       8      3   

   Sixth   T  Debut Since/LastApp BestPosition  goal_diff  
0      4  79   1929          1929            1       2807  

[1 rows x 21 columns]
******************************************

Team with minimum Go

**Observations :**
* **Real Madrid Team** has a Maximum Goal difference of 2807.It means it has done 2807 more than the other teams has done goal to them.
* **Racing Santander** Team has minimum Goal difference of -525. It means other team has done 525 more goals to them than total of goals done by Racing Santander.

### Q5. Create a new column with name “Winning Percent” and append it to the data set .

In [100]:
#creating a new column with "Winning percent"
laliga_df['Winning_percent'] = laliga_df['GamesWon']/laliga_df['GamesPlayed']*100

In [101]:
laliga_df

Pos              Team  Seasons  Points  GamesPlayed  GamesWon  GamesDrawn  \
0     1       Real Madrid       86    4385         2762      1647         552   
1     2         Barcelona       86    4262         2762      1581         573   
2     3   Atletico Madrid       80    3442         2614      1241         598   
3     4          Valencia       82    3386         2664      1187         616   
4     5   Athletic Bilbao       86    3368         2762      1209         633   
..  ...               ...      ...     ...          ...       ...         ...   
56   57             Xerez        1      34           38         8          10   
57   58            Condal        1      22           30         7           8   
58   59   Atletico Tetuan        1      19           30         7           5   
59   60  Cultural Leonesa        1      14           30         5           4   
60   61            Girona        1       0            0         0           0   

    GamesLost  GoalsFor  GoalsAgainst  ...  Third  Fourth  Fifth  Sixth   T  \
0         563      5947          3140  ...      8       8      3      4  79   
1         608      5900          3114  ...     12      12      4      6  83   
2         775      4534          3309  ...     16       9      7      6  56   
3         861      4398          3469  ...     10      11     10      7  50   
4         920      4631          3700  ...     10       5      8     10  49   
..        ...       ...           ...  ...    ...     ...    ...    ...  ..   
56         20        38            66  ...      0       0      0      0   0   
57         15        37            57  ...      0       0      0      0   0   
58         18        51            85  ...      0       0      0      0   0   
59         21        34            65  ...      0       0      0      0   0   
60          0         0             0  ...      0       0      0      0   0   

      Debut  Since/LastApp BestPosition goal_diff  Winning_percent  
0      1929           1929            1      2807        59.630702  
1      1929           1929            1      2786        57.241130  
2      1929        2002-03            1      1225        47.475134  
3   1931-32        1987-88            1       929        44.557057  
4      1929           1929            1       931        43.772629  
..      ...            ...          ...       ...              ...  
56  2009-10        2009-10           20       -28        21.052632  
57  1956-57        1956-57           16       -20        23.333333  
58  1951-52        1951-52           16       -34        23.333333  
59  1955-56        1955-56           15       -31        16.666667  
60  2017-18        2017-18            9         0              NaN  

[61 rows x 22 columns]

In [102]:
#Top 5 Teams with highest winning percent
high_win_percentage = laliga_df.sort_values(['Winning_percent'], ascending=[False])
high_win_percentage.head(5)

Pos             Team  Seasons  Points  GamesPlayed  GamesWon  GamesDrawn  \
0    1      Real Madrid       86    4385         2762      1647         552   
1    2        Barcelona       86    4262         2762      1581         573   
2    3  Atletico Madrid       80    3442         2614      1241         598   
3    4         Valencia       82    3386         2664      1187         616   
4    5  Athletic Bilbao       86    3368         2762      1209         633   

   GamesLost  GoalsFor  GoalsAgainst  ...  Third  Fourth  Fifth  Sixth   T  \
0        563      5947          3140  ...      8       8      3      4  79   
1        608      5900          3114  ...     12      12      4      6  83   
2        775      4534          3309  ...     16       9      7      6  56   
3        861      4398          3469  ...     10      11     10      7  50   
4        920      4631          3700  ...     10       5      8     10  49   

     Debut  Since/LastApp BestPosition goal_diff  Winning_percent  
0     1929           1929            1      2807        59.630702  
1     1929           1929            1      2786        57.241130  
2     1929        2002-03            1      1225        47.475134  
3  1931-32        1987-88            1       929        44.557057  
4     1929           1929            1       931        43.772629  

[5 rows x 22 columns]

**Observations:**
* These are the Top 5 teams with the highest winning percentage.
* **Real Madrid has the highest probability of winning Laliga Tournament.**

### Q6. Group teams based on their “Best position” and print the sum of their points for all positions.

In [103]:
#Grouping based on Bestposition and sum of Points.
print(laliga_df.groupby('BestPosition')['Points'].sum())

BestPosition
1     27933
2      6904
3      5221
4      6563
5      1884
6      2113
7      1186
8      1134
9        96
10      450
11      445
12      511
14       71
15       14
16       81
17      266
19       81
20       34
Name: Points, dtype: int64


**These are the sum of points scored by teams when being at that position.**